### ISSUE TITLE
Decoding WebM files with alpha

### ISSUE THREAD

**Issue by kareljuricka** [w3c/webcodecs#377]
Do webcodec decoding operation support alpha channels in webm files?

I'm using https://github.com/Yahweasel/mkvdemuxjs demuxer for getting chunks and webcodes VideoDecoder for getting VideoFrames. But when I draw frame on canvas, alpha layer from input webm file is missing. 

I'm not sure if it's problem of demuxer or of webcodecs.. 
I noticed that encoder does support alpha: 'keep' option, I looked for something similar for decoder but with no luck

**Comment by StaZhu** [comment]
HEVC with Alpha decoding on macOS has recently been implemented in the latest Chrome Canary 108.

It supports WebCodec API well, meanwhile can preserve its alpha layer.

A simple demo here FYI: `https://lf-tk-sg.ibytedtos.com/obj/tcs-client-sg/resources/webcodec.html?file=hevc_with_alpha_layer.mp4`

**Comment by rcunning** [comment]
@sandersdan any pointers on how to encode the alpha channel as a separate stream and combine in JS? I have the need to alpha video + webcodecs that will work correctly on Win+Mac in Chrome.

**Comment by dalecurtis** [comment]
To combine in JS you'll need the opaque pixel data and alpha data in CPU memory. You can then merge them into a I420A frame or RGB convert into RGBA/BGRA. 

There's no way to zip hardware decoded opaque data with hardware/software decoded alpha data with the current API, so you'll need to use `copyTo()` to retrieve one or both if hardware decoding is used.

**Comment by rcunning** [comment]
> To combine in JS you'll need the opaque pixel data and alpha data in CPU memory. You can then merge them into a I420A frame or RGB convert into RGBA/BGRA.
> 
> There's no way to zip hardware decoded opaque data with hardware/software decoded alpha data with the current API, so you'll need to use `copyTo()` to retrieve one or both if hardware decoding is used.

Thanks for responding so quickly @dalecurtis . So it looks pretty easy to extract out the alpha channel with ffmpeg using the alphaextract filter. And also easy to decode both the regular file + alpha file - modifying the webcodec decode sample (https://github.com/w3c/webcodecs/tree/main/samples/video-decode-display). However a couple things:

1. Assuming I can copy these data out correctly, can you elaborate on how I would merge them to RGBA in JS? Is there any sample code you know of that might give me a starting point? Eventually I want put the RGBA frame into a pixi.js texture (i.e. PIXI.Texture.fromBuffer), although it seems like there would be a way to combine the two in webgl/pixi instead of in JS. 

2. to use copyTo I need the frame's allocation size, but some files do not include a format in their VideoFrames:
```
let buffer = new Uint8Array(frame.allocationSize());
frame.copyTo(buffer).then(layout => {
       ...
})
```
which gives me `Uncaught DOMException: Failed to execute 'allocationSize' on 'VideoFrame': Operation is not supported when format is null.`
I can work around this by specifying the pix_fmt to yuv420p at encode time, but I am unclear why this fails. 


Thank again for your help

**Comment by dalecurtis** [comment]
More recent versions of Chrome should have less cases where the format is null. When it's null you need to go through ImageBitmap or canvas to read back the pixel data. If the data you readback is in RGBx or BGRx format you just need to replace each alpha byte with the corresponding one from your alpha stream.

If your draw primitives can source the alpha plane from elsewhere you don't even need to zip into RGBA.

**Comment by rcunning** [comment]
@dalecurtis sorry to bother you again, but I am back on this task and having an issue with the data coming back from copyTo. It appears to be YUV420p by its size, and its y values seem correct. For a 640x360 frame, layout returned shows `[{ "offset": 0, "stride": 640 }, { "offset": 230400, "stride": 640 }]` with a total size of 345600 -- I would expect the layout to have 3 elements showing u and v as separate offsets to match YUV420p. In any case, I am having trouble converting these data to RGBA because the u and v appear to be not laid out the way I would expect. e.g. a simple rgba JS conversion would pull these out like:
```
for (let i = 0; i < width * height; i++) {
    const y = yuv[i];
    const u = yuv[(i >> 2) + (width * height)];
    const v = yuv[(i >> 2) + (width * height * 5) / 4];
...
```
So my images are coming out with b/w looking correct, but the color incorrect. Is there something I am missing here?

e.g.
<img width="199" alt="image" src="https://user-images.githubusercontent.com/26164/220217450-9c4040bc-97c0-47ef-b56e-78e9d4f3d228.png">

**Comment by dalecurtis** [comment]
Given that you only have two returned layouts, it seems like it's NV12 instead of I420?

**Comment by rcunning** [comment]
> Given that you only have two returned layouts, it seems like it's NV12 instead of I420?

omg, yes. Can't believe I did not notice that. Thank you again.

**Comment by rcunning** [comment]
The issue is closed, but I thought @dalecurtis might want to see the webcodecs-powered alpha video feature which finally was enabled for general use in Scenery this week:

https://twitter.com/rcunning/status/1679993435003015168

**Comment by orange4glace** [comment]
@rcunning Could you explain how did you make h265 alpha channel video? I tried with FFMPEG + libx265 but it failed.

**Comment by rcunning** [comment]
> [@rcunning](https://github.com/rcunning) Could you explain how did you make h265 alpha channel video? I tried with FFMPEG + libx265 but it failed.

I did not use h265! I created am .mp4 with two h264 video tracks:
- yuv420p for the colors
- yuv420p for the alpha channel (so looks like a gray-scale video)

Then at decode+render time I combined each paired VideoFrame rendered textures to get a single RGBA texture for display.

**Comment by orange4glace** [comment]
> I did not use h265! I created an .mp4 with two h264 video tracks:

Thanks for your reply!

So basically you have two `VideoDecoder`, one for colors and another for alpha channel?

**Comment by orange4glace** [comment]
> I did not use h265! I created an .mp4 with two h264 video tracks:

Thanks for your reply @rcunning !

So basically you have two `VideoDecoder`, one for colors and another for alpha channel?

**Comment by rcunning** [comment]
> So basically you have two `VideoDecoder`, one for colors and another for alpha channel?

Yep.

**Comment by yisibl** [comment]
> Hi [@sandersdan](https://github.com/sandersdan), can Chrome encode VP9/VP8 video with alpha now?

See https://issues.chromium.org/issues/370423037

**Comment by dalecurtis** [comment]
No, Chrome still doesn't support this natively. You'll still need to use two encoders as discussed above.
